README.md
#WooCommerce Arrow Payments Plugin

Contributors: seanvfs

WooCommerce Arrow Payments Plugin allows you to accept payments on WooCommerce via the Arrow Payments gateway

####Changelog
12-03-2015 - version 1.0
* Initial Release

####Test Credentials
* API Key: 5cbb15c1d87d480f8b82ee1a3abba353
* Merchant ID: 1231616282
* Test Customer ID: 96722
* Test Payment Method ID: 46356


changelog.txt
*** Arrow Payments Changelog ***

12-03-2015 - version 1.0
* Initial Release

templates/cc-form.php
<div>
    <p class="form-row form-row-wide">
        <label><?php _e('Credit Card Number', 'wc-arrowpayments'); ?></label>
        <input type="text" class="input-text" size="15" name="billing-cc-number" value="<?php echo ($prefill ? '4111111111111111' : ''); ?>"></input>
    </p>
    <p class="form-row form-row-first">
        <label><?php _e('Exp. (mm/yy)', 'wc-arrowpayments'); ?></label>
        <input type="text" class="input-text" size="4" name="billing-cc-exp" value="<?php echo ($prefill ? '11/17' : ''); ?>"></input>
    </p>
    <p class="form-row form-row-last">
        <label><?php _e('CVV', 'wc-arrowpayments'); ?></label>
        <input type="text" class="input-text" size="4" name="billing-cvv" value="<?php echo ($prefill ? '111' : ''); ?>"></input>
    </p>
</div>
<input type="submit" value="<?php _e('Confirm and pay', 'wc-arrowpayments'); ?>" class="submit buy button">

arrowsdk/class-arrowpayments-sdk.php
<?php

class WC_Arrow_Payments_SDK {
	/*
		REST Post Client Helper
	*/
	public static function restcall($url,$vars,$ssl) {
		$headers = array(
			'Accept: application/json',
			'Content-Type: application/json',
		);
		$data = json_encode( $vars );
		$handle = curl_init();
		curl_setopt($handle, CURLOPT_URL, $url);
		curl_setopt($handle, CURLOPT_HTTPHEADER, $headers);
		curl_setopt($handle, CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($handle, CURLOPT_SSL_VERIFYHOST, $ssl);
		curl_setopt($handle, CURLOPT_SSL_VERIFYPEER, $ssl);
		curl_setopt($handle, CURLOPT_POST, 1);
		curl_setopt($handle, CURLOPT_POSTFIELDS, $data);
		curl_setopt($handle, CURLOPT_HEADER, true);
		$response = curl_exec($handle);
		$code = curl_getinfo($handle, CURLINFO_HTTP_CODE);
		$header_size = curl_getinfo($handle, CURLINFO_HEADER_SIZE);
		$header = substr($response, 0, $header_size);
		$body = substr($response, $header_size);
		$full_response = array($body, $header);
		// Check for errors
		if($response === FALSE){
			die(curl_error($handle));
		}
		return $full_response;
	}	
	
	/*
		String Helper for error message
	*/
	public static function get_string_between($string, $start, $end){
		$string = " ".$string;
		$ini = strpos($string,$start);
		if ($ini == 0) return "";
		$ini += strlen($start);
		$len = strpos($string,$end,$ini) - $ini;
		return substr($string,$ini,$len);
	}
	/**
     * A snippet to send to ArrowPayments to redirect the user back to the
     * merchant's server. Use this on your relay response page.
     *
     * @param string $redirect_url Where to redirect the user.
     *
     * @return string
     */
    public static function getRelayResponseSnippet($redirect_url)
    {
        return "<html><head><script language=\"javascript\">
                <!--
                window.location=\"{$redirect_url}\";
                //-->
                </script>
                </head><body><noscript><meta http-equiv=\"refresh\" content=\"1;url={$redirect_url}\"></noscript></body></html>";
    }
	
	/*
		Post the transaction form for processing by the Arrow API
	*/
	public static function Step1($amount, $order, $api_key, $mid, $relay_response_url, $testmode = true, $test_customer_id = '96722', $test_payment_method_id = '46356')
	{		
		// Handle gateway negotiation
		if($testmode == true){
			$gateway_url = 'http://demo.arrowpayments.com/api/transaction/start';
			$ssl = 0;
		}else{
			$gateway_url = 'https://gateway.arrowpayments.com/api/transaction/start';
			//Temporarily changed to 0 from 2 so we're not validating the SSL on Arrow's Gateway
			$ssl = 0;
		}		
		// Compose request to start transaction (Level 2 Transaction)
		$data = array(
					"ApiKey" => $api_key, 
					"MID" => $mid,
					"ReturnUrl" => $relay_response_url,
					"TransactionType" => 'sale',
					"Amount" => $amount,
					"TaxAmount" => $order->get_total_tax(),
					"ShippingAmount" => $order->get_total_shipping(),
					"CustomerPONumber" => $order->order_key,
					"Description" => $order->get_order_number(),
					"CustomerId" => 0,
					"Customer" => array(
											"Name" => $order->billing_first_name . ' ' . $order->billing_last_name . '-' . date('yMd-Hms'),
											"Code" => date('yMd-Hms') . rand()
										),
					"SaveNewPaymentMethod" => false,
					"BillingAddress" => array(
												"Address1" => $order->billing_address_1,
												"Address2" => $order->billing_address_2,
												"City" => $order->billing_city,
												"State" => $order->billing_state,
												"Postal" => $order->billing_postcode,
												"Phone" => $order->billing_phone,
											)					
					);				
		// Perform REST call to Arrow Payments
		try {
			$result = self::restcall($gateway_url, $data, $ssl);
		}
		catch (Exception $e) {
			echo 'Caught exception: ', $e->getMessage(), '\n';
		}		
		// Return response
		return $result[0];
	}		
	/*
		Prepare and present payment form to customer
	*/
	public static function Step2($step1JSON, $testmode)
    {
		// Carry session data through
		if($testmode == true){
			$prefill = true;
		}else {
			$prefill = false;
		}		
		// Decode JSON response from step 1
		$json = json_decode($step1JSON);
		
		// Arrow Payment POST Url
        $post_url = ( $json->FormPostUrl );	 
		
		$hidden_fields = '';
		
		ob_start();

        woocommerce_get_template( 'cc-form.php', array(
			'prefill' => $prefill
		), 'arrowpayments/', plugin_dir_path( __DIR__ ) . 'templates/' );

		$form = '<form method="post" action="' . $post_url . '"> ' . $hidden_fields;
        $form .= ob_get_clean();
        $form .= '</form>';
        return $form;
    }
	/*
		Complete Transaction and handle WooCommerce Order Updates
	*/
	public static function Step3($api_key, $testmode)
	{
		global $woocommerce;
		// Get the token id from return url from Step2
		$token_id = $_GET['token-id'];

		// Handle gateway negotiation
		if($testmode == true){
			$gateway_url = 'http://demo.arrowpayments.com/api/transaction/complete';
			$ssl = 0;
		}else{
			$gateway_url = 'https://gateway.arrowpayments.com/api/transaction/complete';
			//Temporarily changed to 0 from 2 so we're not validating the SSL on Arrow's Gateway
			$ssl = 0;
		}
		// Compose request to complete transaction
		$data = array(
						"ApiKey" => $api_key,
						"TokenId" => $token_id
					);
		// Perform REST call to Arrow Payments
		try {
			$result = self::restcall($gateway_url, $data, $ssl);
		}
		catch (Exception $e) {
			echo 'Caught exception: ', $e->getMessage(), '\n';
		}			
		if($result[0] == 'Bad Request')
		{
			$errorMsg = self::get_string_between($result[1], "Error: ", " REFID");
		}
		else
		{
			// Decode response
			$json = json_decode($result[0]);
		}	
		// Get the order number - DONE
		$orderNum = str_replace('#','',$json->Description);
		
		// Get the order object - DONE
		$order = new WC_Order( (int) $orderNum);
		
		// Get redirect url - DONE
		$redirect_url = WC_Gateway_Arrow_Payments::get_return_url( $order );
		
		if ( $json->Success == 1 ) {
			if ( $order->key_is_valid( $json->CustomerPONumber )) {

				// Payment complete
				$order->payment_complete();

				// Redirect URL
				$redirect_url = add_query_arg( 'response_code', 1, $redirect_url );
				$redirect_url = add_query_arg( 'transaction_id', $json->ID, $redirect_url );

				// Remove cart
				$woocommerce->cart->empty_cart();
			}
			else 
			{
				// Key did not match order id
				$order->add_order_note( sprintf(__('Payment received, but order ID did not match key: Success: %s - %s.', 'wc-arrowpayments'), $json->Success, $json->Message ) );

				// Put on hold if pending
				if ($order->status == 'pending' || $order->status == 'failed') {
					$order->update_status( 'on-hold' );
				}
			}
		} 
		else 
		{
			// Mark failed
			$order->update_status( 'failed', sprintf(__('Payment failure: Success: %s - %s.', 'wc-arrowpayments'), $json->Success, $json->Message ) );

			$redirect_url = $order->get_checkout_payment_url( true );
			$cleanError = str_replace(array("\r\n", "\r", "\n"), "",preg_replace('/[^a-zA-Z0-9\s]/', '', strip_tags(html_entity_decode($errorMsg))));
			$redirect_url = add_query_arg( 'wc_error', __( 'Error: ', 'wc-arrowpayments' ) . ' '.$cleanError, $redirect_url );

			if ( is_ssl() || get_option( 'woocommerce_force_ssl_checkout' ) == 'yes' ) 
				$redirect_url = str_replace( 'http:', 'https:', $redirect_url );
		}
		echo self::getRelayResponseSnippet($redirect_url);
	}	
}


gateway-arrowpayments.php
<?php
/*
Plugin Name: WooCommerce Arrow Payments Gateway
Plugin URI: http://www.visionfriendly.com
Description: Extends WooCommerce with an <a href="https://www.arrowpayments.com" target="_blank">Arrow Payments</a> gateway. An Arrow Payments merchant account is required.
Version: 1.0
Author: Sean Herbert
Author URI: http://www.visionfriendly.com/
Copyright: © 2015 VisionFriendly.com.
License: GNU General Public License v3.0
License URI: http://www.gnu.org/licenses/gpl-3.0.html
API Docs: http://gateway.arrowpayments.com/apidocumentation?/api
*/
add_action('plugins_loaded', 'woocommerce_arrowpayments_init', 0);
function woocommerce_arrowpayments_init() {
	if ( ! class_exists( 'WC_Payment_Gateway' ) ) return;
    /**
     * Localisation
     */
    load_plugin_textdomain('wc-arrowpayments', false, dirname( plugin_basename( __FILE__ ) ) . '/languages');
	/**
 	 * Gateway class
 	 */
	class WC_Gateway_Arrow_Payments extends WC_Payment_Gateway {
		public function __construct() {
			global $woocommerce;
			$this->id					= 'arrowpayments';
			$this->method_title 		= __('Arrow Payments', 'wc-arrowpayments');
			$this->method_description 	= __('Arrow Payments handles all the steps in the secure transaction while remaining virtually transparent. Payment data is passed from the checkout to Arrow Payments for processing thus simplifying PCI compliance.', 'wc-arrowpayments');
			$this->icon 				= WP_PLUGIN_URL . "/" . plugin_basename( dirname(__FILE__)) . '/images/cards.png';
			// Load the form fields
			$this->init_form_fields();
			// Load the settings.
			$this->init_settings();
			// Get setting values
			$this->title           				= $this->settings['title'];
			$this->description     				= $this->settings['description'];
			$this->enabled         				= $this->settings['enabled'];
			$this->api_key         				= $this->settings['api_key'];
			$this->mid 			   				= $this->settings['mid'];
			$this->testmode        				= isset( $this->settings['testmode'] ) && $this->settings['testmode'] == 'no' ? false : true;
			$this->test_customer_id     		= $this->settings['test_customer_id'];
			$this->test_payment_method_id       = $this->settings['test_payment_method_id'];
			// Payment form hook
			add_action( 'woocommerce_receipt_arrowpayments', array( $this, 'receipt_page' ) );
			// Payment listener/API hook
			$this->relay_response_url = $woocommerce->api_request_url( get_class() );
			add_action( 'woocommerce_api_wc_gateway_arrow_payments', array( $this, 'relay_response' ) );
			add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
		}
		/**
	     * Initialise Gateway Settings Form Fields
	     */
	    public function init_form_fields() {
	    	$this->form_fields = array(
				'enabled' => array(
					'title'       => __( 'Enable/Disable', 'wc-arrowpayments' ),
					'label'       => __( 'Enable Arrow Payments', 'wc-arrowpayments' ),
					'type'        => 'checkbox',
					'description' => '',
					'default'     => 'no'
				),
				'title' => array(
					'title'       => __( 'Title', 'wc-arrowpayments' ),
					'type'        => 'text',
					'description' => __( 'This controls the title which the user sees during checkout.', 'wc-arrowpayments' ),
					'default'     => __( 'Credit card', 'wc-arrowpayments' )
				),
				'description' => array(
					'title'       => __( 'Description', 'wc-arrowpayments' ),
					'type'        => 'textarea',
					'description' => __( 'This controls the description which the user sees during checkout.', 'wc-arrowpayments' ),
					'default'     => 'Pay securely using your credit card.'
				),
				'api_key' => array(
					'title'       => __( 'API Key', 'wc-arrowpayments' ),
					'type'        => 'text',
					'description' => __( 'Look this up by logging into your Arrow Payments account.', 'wc-arrowpayments' ),
					'default'     => '5cbb15c1d87d480f8b82ee1a3abba353'
				),
				'mid' => array(
					'title'       => __( 'Merchant ID', 'wc-arrowpayments' ),
					'type'        => 'text',
					'description' => __( 'Look this up by logging into your Arrow Payments account.', 'wc-arrowpayments' ),
					'default'     => '1231616282'
				),
				'testmode' => array(
					'title'       => __( 'Use Test Environment', 'wc-arrowpayments' ),
					'label'       => __( 'Yes, I am using a the test environment.', 'wc-arrowpayments' ),
					'type'        => 'checkbox',
					'description' => __( 'Enable this if the above keys are for a developer/test account used for testing. If you are using your live account, do not enable this setting.', 'wc-arrowpayments' ),
					'default'     => 'yes'
				),
				'test_customer_id' => array(
					'title'       => __( 'Test Customer ID', 'wc-arrowpayments' ),
					'type'        => 'text',
					'description' => __( '', 'wc-arrowpayments' ),
					'default'     => '96722'
				),				
				'test_payment_method_id' => array(
					'title'       => __( 'Test Payment Method ID', 'wc-arrowpayments' ),
					'type'        => 'text',
					'description' => __( '', 'wc-arrowpayments' ),
					'default'     => '46356'
				),
			);
	    }
		/**
	     * Check if this gateway is enabled
	     */
		public function is_available() {
			if ( ! $this->api_key || ! $this->mid ) 
				return false;

			return parent::is_available();
		}
		/**
		 * Process the payment and return the result - this will redirect the customer to the pay page
		 */
		public function process_payment( $order_id ) {

			$order = new WC_Order( $order_id );

			return array(
				'result' 	=> 'success',
				'redirect'	=> $order->get_checkout_payment_url( true )
			);
		}
		/**
		 * Receipt_page for showing the payment form which sends data to Arrow Payments
		 */
		public function receipt_page( $order_id ) {
			global $woocommerce;

			// Include the SDK
			if ( ! class_exists( 'WC_Arrow_Payments_SDK' ) )
				require_once( dirname(__FILE__) . '/arrowsdk/class-arrowpayments-sdk.php');

			// Get the order
			$order = new WC_Order( $order_id );

			// Get the amount
			$amount = $order->get_total();
			
			// Test Print out $order object
			//echo '<pre>';
			//print_r($order);
			//echo '</pre>';

			echo wpautop(__('Enter your payment details below and click "Confirm and pay" to securely pay for your order.', 'wc-arrowpayments'));
			
			// Set up new WC_Arrow_Payments_SDK object
			$sdk = new WC_Arrow_Payments_SDK();
			
			// Step 1 - Get Form URL from Arrow Payments
			$step1Result = $sdk->Step1($amount, $order, $this->api_key, $this->mid, $this->relay_response_url, $this->testmode, $this->test_customer_id, $this->test_payment_method_id);

			// Step 2 - Show the payment form
			if($step1Result != 'Bad Request'){
				$step2Result = $sdk->Step2($step1Result, $this->testmode);
				echo $step2Result;
			}
			else
			{
				$error = new WP_Error( 'arrowpayments_error', __( 'Sorry, there was a problem communicating with the gateway.  Please try again.', 'wc-arrowpayments' ) );
				echo $error->get_error_message();
			}
		}
		/**
		 * Relay response - handles return data from Arrow Payments and does redirects
		 */
		public function relay_response() {
			global $woocommerce;
			
			// Clean
			@ob_clean();
			
			// Header
			header( 'HTTP/1.1 200 OK' );
			
			// Include the SDK
			if ( ! class_exists( 'WC_Arrow_Payments_SDK' ) )
				require_once( dirname(__FILE__) . '/arrowsdk/class-arrowpayments-sdk.php');	
				
			// Set up new WC_Arrow_Payments_SDK object
			$sdk = new WC_Arrow_Payments_SDK();
			
			// Process Response
			echo $sdk->Step3($this->api_key, $this->testmode);
			
			exit;
		}
	} // end WC_Gateway_Arrow_Payments
	/**
 	* Add the Gateway to WooCommerce
 	**/
	function woocommerce_add_arrowpayments_gateway($methods) {
		$methods[] = 'WC_Gateway_Arrow_Payments';
		return $methods;
	}
	add_filter('woocommerce_payment_gateways', 'woocommerce_add_arrowpayments_gateway' );
}